server/src/controllers/favourite.controller.ts

```diff
@@ -0,0 +1,87 @@
+import { Request, Response, NextFunction, Router } from 'express';
+import Controller from '../interfaces/controller.interface';
+import userModel from '../models/user.model';
+import restaurantModel from '../models/restaurant.model';
+import RestaurantAlreadyInFavourites from '../exceptions/RestaurantAlreadyInFavourites';
+import RestaurantNotFoundException from '../exceptions/RestaurantNotFoundException';
+import UserNotFoundException from '../exceptions/UserNotFoundException';
+import RestaurantIsNotonTheList from '../exceptions/RestaurantIsNotOnTheList';
+import permissionMiddleware from '../middleware/permission.middleware';
+
+class FavouriteController implements Controller {
+  public path = '/favourites';
+  public router = Router();
+  private restaurant = restaurantModel;
+  private user = userModel;
+
+  constructor() {
+    this.initializeRoutes();
+  }
+
+  private initializeRoutes() {
+    this.router.get(`${this.path}/:userId`, permissionMiddleware, this.getFavourites);
+    this.router.put(`${this.path}/:userId`, permissionMiddleware, this.addOrRemoveRestaurantToFavourites);
+  }
+
+  private addOrRemoveRestaurantToFavourites = async (request: Request, response: Response, next: NextFunction) => {
+    const userId: string = request.params.userId;
+    if (!userId.match(/^[0-9a-fA-F]{24}$/)) {
+      next(new UserNotFoundException(userId))
+    }
+
+    const user = await this.user.findById(userId);
+    if (!user) {
+      next(new UserNotFoundException(userId))
+    }
+
+    const restaurantId: string = request.body.restaurantId;
+    if (!restaurantId.match(/^[0-9a-fA-F]{24}$/)) {
+      next(new RestaurantNotFoundException(restaurantId));
+    }
+
+    const restaurant = await this.restaurant.findById(restaurantId);
+    if (!restaurant) {
+      next(new RestaurantNotFoundException(restaurantId));
+    }
+
+    const favourites = user.get('favourites', null, {getters: false});
+    const operation: string = request.body.operation;
+
+    if (operation === "add") {
+      if (favourites.includes(restaurantId)) {
+      next(new RestaurantAlreadyInFavourites(restaurantId));
+      } else {
+        favourites.push(restaurantId);
+        await user.save();
+        response.send(restaurant);
+      }
+    }
+    if (operation === "remove") {
+      if (!favourites.includes(restaurantId)) {
+        next(new RestaurantIsNotonTheList(restaurantId));
+      } else {
+        const index = favourites.findIndex((el:any) => el._id === restaurantId);
+        favourites.splice(index, 1);
+        await user.save();
+        response.send(200);
```

Comment:
Z tego co wiem to domyślna wartość statusu dla response jest 200 więc można ten zapis pominąć.